internal/index/indexscanner/scanner.go

```diff
@@ -98,26 +98,50 @@ func ReadPluginFromFile(path string) (index.Plugin, error) {
 
 func ReadPlugin(f io.ReadCloser) (index.Plugin, error) {
 	defer f.Close()
-	p, err := DecodePluginFile(f)
+	plugin := &index.Plugin{}
+	err := decodeFile(f, plugin)
+	p := *plugin
 	if err != nil {
 		return p, errors.Wrap(err, "failed to decode plugin manifest")
 	}
 	return p, errors.Wrap(validation.ValidatePlugin(p.Name, p), "plugin manifest validation error")
 }
 
-// DecodePluginFile tries to decodes a plugin manifest from r.
-func DecodePluginFile(r io.Reader) (index.Plugin, error) {
-	var plugin index.Plugin
+// ReadReceiptFromFile loads a file from the FS. When receipt file not found, it
+// returns an error that can be checked with os.IsNotExist.
+func ReadReceiptFromFile(path string) (index.Receipt, error) {
+	f, err := os.Open(path)
+	if os.IsNotExist(err) {
+		// TODO(ahmetb): we should use go1.13+ errors.Is construct at call sites to evaluate if an error is os.IsNotExist
+		return index.Receipt{}, err
+	} else if err != nil {
+		return index.Receipt{}, errors.Wrap(err, "failed to open index file")
+	}
+	return ReadReceipt(f)
```

Comment:
This is pure duplication.

Please implement something like `decodeFile()` that works for both Receipt and Plugin to deduplicate implementation. This should be just:

```
func ReadReceiptFromFile(f file) (Receipt, error) {
    var v Receipt
    return decodeFile(f, &v)
}

Reply:
Actually I've just noticed you've done ReadReceiptFromFile &rarr; ReadReceipt &rarr; decodeFile.

However, the core logic seems still very much copypasta from the implementation for Plugin type. So let's try to minimize those if they both work the same way.

Reply:
Ya the main problem is that `ReadPluginFromFile`/`ReadReceiptFromFile` return a concrete type `Plugin`/`Receipt`; both of them just read a file and unmarshal to one or the other. I minimized the duplication as much as I could but you're right that it is pretty much a copypasta of the plugin implementation. I'll take another pass at it to see what else I can do, like moving the `ReadReceipt` function into `ReadReceiptFromFile`.